### ISSUE TITLE
Mixed language encoding wrong in Outlook

### ISSUE THREAD

**Issue by chrislim1990** [meitar/wp-pgp-encrypted-emails#34]
Original report:  
https://wordpress.org/support/topic/multi-language-email-mail-body-encoding-are-wrong-on-outlook/

### Issue:
When send a encrypted mixed language to Outlook, the body failed to apply the correct encoding. The encoding type used is SMIME.

### Platform:
Tested on Windows 10's Outlook and Outlook for Mac 16.15.

### Steps to reproduce:
1. Add some non-english text to the test message body text from plugin. This is my test message "你好 Allô! This is a test message".
2. Add SMIME to the user and click "Send me a test mail".
3. Check the mail in Outlook, the non-english text are not display. Attached with the screenshot and raw source. <br>
Notice the weird character in front of English text:
![wrong encoding](https://user-images.githubusercontent.com/3245469/44147150-6d3c5420-a0c5-11e8-9f18-a619de0f3516.png)<br>
The expected result:
![correct encoding](https://user-images.githubusercontent.com/3245469/44147205-b3ae08d6-a0c5-11e8-9506-fc39159195a3.png)<br>
Here is the raw mail source:
[raw source.txt](https://github.com/meitar/wp-pgp-encrypted-emails/files/2290123/raw.source.txt)

### Supporting Information:
1. If send an mixed language mail directly from client to client, eg Mac Mail to Outlook, the encoding do displayed correctly. Enclosed with the client to client mail's raw source.
[raw source - client to client mail.txt](https://github.com/meitar/wp-pgp-encrypted-emails/files/2290130/raw.source.-.client.to.client.mail.txt)
2. If the mail is unencrypted, Outlook do display correct encoding.
3. If change the charset via wp_mail_charset filter, only encrypted mail subject reflected the changed charset, mail body do not show any changes at all. <br>
Notice that the body encoding still the same although I have change the charset:
![charset changed](https://user-images.githubusercontent.com/3245469/44147506-fa6a20d8-a0c6-11e8-8952-c09266495cfa.png)<br>
This is the changed charset:
![charset](https://user-images.githubusercontent.com/3245469/44147618-5d8462be-a0c7-11e8-8c99-2d72760ed40a.png)<br>
The raw source for charset changed email:
[raw source - charset changed.txt](https://github.com/meitar/wp-pgp-encrypted-emails/files/2290147/raw.source.-.charset.changed.txt)

4. If I added utf8_decode to the mail body in the sendTestEmail function, the "ô" in Allô do displayed correctly in Outlook, but the Chinese character get converted into question mark.<br>
Notice that the "ô" in Allô is displayed correctly:
![screen shot 2018-08-15 at 8 43 45 pm](https://user-images.githubusercontent.com/3245469/44148782-3d7c89b0-a0cc-11e8-9c8b-2e35af551794.png)

**Comment by fabacab** [comment]
Hi @chrislim1990, thank you for this report. I have pushed f79b98fb2ae1a53b790b1c93c8c4cd89e1701f7c to the `pkcs7` testing branch. Would you please test your issue against that new version of the code and report back here regarding the result? ([Here is a zipfile you can use to manually install the plugin with this patch](https://github.com/fabacab/wp-pgp-encrypted-emails/archive/f79b98fb2ae1a53b790b1c93c8c4cd89e1701f7c.zip).)

**Comment by chrislim1990** [comment]
Thank you very much for the prompt update! I have installed the new code, the complete 
content type header did show up correctly.

However, the encoding still gibberish in Outlook. In fact, the gibberish character looks exactly the same with the previous one.

Here is the outcome in Outlook:
![screen shot 2018-08-16 at 12 35 49 pm](https://user-images.githubusercontent.com/3245469/44188587-eec2ce80-a150-11e8-9ad3-108435562df2.png)

--- 

I did a quick test on the mail source, by replacing the server sent mail's header with client to client mail's header, the encoding still gibberish. But if I replace the mail body instead, the encoding will display correctly. 

Below is the mail that sent from server, with its mail body replaced by content of client to client mail.
The left panel is the outcome, rendered on Outlook, while the right panel is the raw source. And I notice that the body encoding is rendered correctly:
![screen shot 2018-08-16 at 12 27 33 pm](https://user-images.githubusercontent.com/3245469/44188713-a22bc300-a151-11e8-9be2-f628b0c835c9.png)

----

Really appreciate prompt updates, please let me know if there are anything I can help. Thank you!

**Comment by fabacab** [comment]
Hmm. It sounds like the issue is actually that the content itself is getting mangled somewhere along the way, or perhaps not being treated as UTF-8 input at some point in the pipeline. I'll have to return to this when I have a bit more time. In the mean time, it would help if you can send me an as-complete-as-you-can-report it readout of your WordPress installation so that I can be ready to recreate a testing environment that matches your conditions.

Not sure if I can continue to provide quite so prompt replies for the next little while but now that the issue is documented here (thanks again) I will get back to it eventually.

**Comment by fabacab** [comment]
@chrislim1990 Just wanted to make sure you saw [this note](https://github.com/meitar/wp-pgp-encrypted-emails/pull/31#issuecomment-413667685):

> I had similar issues with non-ascii characters, which I fixed by adding another content type header containing `charset=utf-8` to the "custom header field" in Contact Form 7. The important thing is, that the header with the correct content type and charset must be in the encrypted message body. The correct display of the message *after decryption* therefore depends on the headers specified by the plugin or code sending the email in the first place and is not modified by encrypting it.

In other words, make sure that the code or plugin you are using to generate the e-mail from within WordPress correctly sets its own MIME `Content-Type` header with a full media type and its parameters.

**Comment by chrislim1990** [comment]
Thanks , I confirmed that after adding the content type in custom header field. The foreign text is **displayed correctly** in both Windows and Mac Outlook.

Here is the content type that I have added:
`Content-Type: text/plain; charset=UTF-8`

Just to let you know, if without using any additional form plugin, the body text of test mail will still in gibberish if it contains foreign character. As additional content type for the test mail body are not set.

Anyway, this issue can be marked as resolved and I really appreciate for your time and effort in helping me to solve this issue. Thank you!
